src/scorer.ts

```diff
@@ -7,42 +7,47 @@ export class Scorer {
   p2Cards: CardInstance[];
   p3Cards: CardInstance[];
   p4Cards: CardInstance[];
+  playersCards: Array<Array<CardInstance>>
 
   constructor(p1Cards: string, p2Cards: string, p3Cards: string, p4Cards: string) {
-    this.p1Cards = p1Cards.split(',').filter(a => a).map((name) => getCard(name));
-    this.p2Cards = p2Cards.split(',').filter(a => a).map((name) => getCard(name));
-    this.p3Cards = p3Cards.split(',').filter(a => a).map((name) => getCard(name));
-    this.p4Cards = p4Cards.split(',').filter(a => a).map((name) => getCard(name));
+    this.playersCards = [
+      this.p1Cards = this.buildCardInsts(p1Cards),
+      this.p2Cards = this.buildCardInsts(p2Cards),
+      this.p3Cards = this.buildCardInsts(p3Cards),
+      this.p4Cards = this.buildCardInsts(p4Cards)
+      ]
+
   }
 
   // Functions
-  scoreA(): number[] {
-    return [
-      this.p1Cards.reduce((total, inst) => total + inst.card.pointsA, 0),
-      this.p2Cards.reduce((total, inst) => total + inst.card.pointsA, 0),
-      this.p3Cards.reduce((total, inst) => total + inst.card.pointsA, 0),
-      this.p4Cards.reduce((total, inst) => total + inst.card.pointsA, 0),
-    ]
-  }
+  scores(): number[] {
+    this.playersCards.forEach((playerCards) => {
+      playerCards.forEach((inst: CardInstance) => {
+        if(typeof inst.card.pointsA === 'function') {
+          inst.finalA = inst.card.pointsA(inst)
+        } else {
+          inst.finalA = inst.card.pointsA
+        }
+        inst.card.calcB?.()
+        inst.card.calcC?.(this.playersCards)
+      })
+    })
 
-  scoreB(): number[] {
-    return [0, 0, 0, 0]
-  }
+    const result: number[] = this.playersCards.map((playerCards) => {
+      return playerCards.reduce((sum, inst: CardInstance) => sum + inst.finalA + inst.finalB || 0, 0)
+    })
 
-  scoreC(): number[] {
-    return [0, 0, 0, 0]
+    return result
   }
 
-  scores(): number[] {
-    const roundAScore: number[] = this.scoreA()
-    const roundBScore: number[] = this.scoreB()
-    const roundCScore: number[] = this.scoreC()
-
-    return [
-      roundAScore[0] + roundBScore[0] + roundCScore[0],
-      roundAScore[1] + roundBScore[1] + roundCScore[1],
-      roundAScore[2] + roundBScore[2] + roundCScore[2],
-      roundAScore[3] + roundBScore[3] + roundCScore[3]
-    ]
+  buildCardInsts(cards: string): CardInstance[] {
+    return cards.split(',').filter(a => a).map((name) => {
+      let realName: string, metadata: string[]
+      [realName, ...metadata] = name.split('+')
+      let metadataValues: Array<string[]> = metadata.map((str) => str.split('=', 2))
```

Comment:
Just use JSON here. This weird string encoding is kind of nonsense.

Reply:
so will pass the data in as objects, but assume it was passed through as JSON from the FE